apps/prairielearn/src/middlewares/autoEnroll.ts

```diff
@@ -8,13 +9,24 @@ export default asyncHandler(async (req, res, next) => {
   // they were enrolled, automatically enroll them. However, we will not
   // attempt to enroll them if they are an instructor (that is, if they have
   // a specific role in the course or course instance) or if they are
-  // impersonating another user.
+  // impersonating another user, or if self-enrollment requires a secret link.
+
+  const courseInstance: CourseInstance = res.locals.course_instance;
+
+  // If we have self-enrollment enabled, and it is before the enabled before date,
+  // then we can enroll the user.
+  const selfEnrollmentEnabled =
```

Comment:
This should always succeed: We don't let users configure self-enrollment, and the default for `self_enrollment_enabled` is `true`, and the default for `self_enrollment_enabled_before_date` is `null`.